File: frontend/src/core/api/billing.ts
/**
 * Billing API client for Stripe integration
 */
import { authConfig } from '../config/auth'

export interface Subscription {
  id: string
  user_id: string
  tier: 'FREE' | 'PRO'
  subscription_type: 'MONTHLY' | 'YEARLY'
  status: 'ACTIVE' | 'INACTIVE' | 'CANCELLED' | 'PAST_DUE' | 'UNPAID'
  current_period_start?: string
  current_period_end?: string
  created_at?: string
  cancelled_at?: string
}

export interface Payment {
  id: string
  user_id: string
  subscription_id?: string
  amount: number
  currency: string
  status: 'PENDING' | 'SUCCEEDED' | 'FAILED' | 'CANCELLED' | 'REFUNDED'
  description?: string
  created_at?: string
  processed_at?: string
}

export interface CheckoutSessionRequest {
  tier: 'PRO'
  subscription_type: 'MONTHLY' | 'YEARLY'
  success_url: string
  cancel_url: string
}

export interface CheckoutSessionResponse {
  checkout_url: string
  session_id: string
}

export interface BillingPortalResponse {
  portal_url: string
}

class BillingApiClient {
  private baseUrl: string

  constructor() {
    this.baseUrl = authConfig.apiUrl
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {},
    accessToken?: string
  ): Promise<T> {
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      ...options.headers,
    }

    if (accessToken) {
      headers['Authorization'] = `Bearer ${accessToken}`
    }

    const response = await fetch(`${this.baseUrl}${endpoint}`, {
      ...options,
      headers,
    })

    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: 'Unknown error' }))
      throw new Error(error.detail || `Request failed: ${response.statusText}`)
    }

    return response.json()
  }

  async createCheckoutSession(
    request: CheckoutSessionRequest,
    accessToken: string
  ): Promise<CheckoutSessionResponse> {
    return this.request<CheckoutSessionResponse>(
      '/api/billing/checkout',
      {
        method: 'POST',
        body: JSON.stringify(request),
      },
      accessToken
    )
  }

  async createBillingPortalSession(
    returnUrl: string,
    accessToken: string
  ): Promise<BillingPortalResponse> {
    const searchParams = new URLSearchParams({ return_url: returnUrl })
    
    return this.request<BillingPortalResponse>(
      `/api/billing/portal?${searchParams}`,
      {
        method: 'POST',
      },
      accessToken
    )
  }

  async getSubscription(accessToken: string): Promise<Subscription> {
    return this.request<Subscription>('/api/billing/subscription', {}, accessToken)
  }

  async cancelSubscription(accessToken: string): Promise<{ message: string }> {
    return this.request<{ message: string }>(
      '/api/billing/subscription',
      {
        method: 'DELETE',
      },
      accessToken
    )
  }

  async getPayments(accessToken: string, limit: number = 50): Promise<Payment[]> {
    const searchParams = new URLSearchParams({ limit: limit.toString() })
    
    return this.request<Payment[]>(
      `/api/billing/payments?${searchParams}`,
      {},
      accessToken
    )
  }

  redirectToCheckout(checkoutUrl: string) {
    window.location.href = checkoutUrl
  }

  redirectToBillingPortal(portalUrl: string) {
    window.location.href = portalUrl
  }
}

export const billingApi = new BillingApiClient()

File: api/app/models/schemas/billing.py
"""Billing related schemas"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

from ...enums import (
    PaymentStatusEnum,
    SubscriptionStatusEnum,
    SubscriptionTypeEnum,
    TierEnum,
)


class SubscriptionCreate(BaseModel):
    """Request model for creating a subscription"""

    tier: TierEnum
    subscription_type: SubscriptionTypeEnum = Field(
        default=SubscriptionTypeEnum.monthly
    )


class SubscriptionResponse(BaseModel):
    """Response model for subscription operations"""

    id: str
    user_id: str
    tier: TierEnum
    subscription_type: SubscriptionTypeEnum
    status: SubscriptionStatusEnum
    current_period_start: Optional[datetime] = None
    current_period_end: Optional[datetime] = None
    created_at: Optional[datetime] = None
    cancelled_at: Optional[datetime] = None


class PaymentCreate(BaseModel):
    """Request model for creating a payment"""

    amount: int = Field(..., gt=0, description="Amount in cents")
    currency: str = Field(default="USD", min_length=3, max_length=3)
    description: Optional[str] = None


class PaymentResponse(BaseModel):
    """Response model for payment operations"""

    id: str
    user_id: str
    subscription_id: Optional[str] = None
    amount: int
    currency: str
    status: PaymentStatusEnum
    description: Optional[str] = None
    created_at: Optional[datetime] = None
    processed_at: Optional[datetime] = None


class CheckoutSessionCreate(BaseModel):
    """Request model for creating a Stripe checkout session"""

    tier: TierEnum
    subscription_type: SubscriptionTypeEnum = Field(
        default=SubscriptionTypeEnum.monthly
    )
    success_url: str
    cancel_url: str


class CheckoutSessionResponse(BaseModel):
    """Response model for checkout session creation"""

    checkout_url: str
    session_id: str


class BillingPortalResponse(BaseModel):
    """Response model for billing portal URL"""

    portal_url: str
